### PR TITLE
8266313: (JEP-356) - RandomGenerator spec implementation requirements tightly coupled to JDK internal classes

### PR DESCRIPTION
**Pull request by JimLaskey**
The wording of the @implSpec referred to internal methods in the description. The patch rewords the @implSpec to be more descriptive of the algorithm than the methods used.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266313](https://bugs.openjdk.java.net/browse/JDK-8266313): (JEP-356) - RandomGenerator spec implementation requirements tightly coupled to JDK internal classes

### Reviewers
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/151/head:pull/151` \
`$ git checkout pull/151`

Update a local copy of the PR: \
`$ git checkout pull/151` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/151/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 151`

View PR using the GUI difftool: \
`$ git pr show -t 151`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/151.diff">https://git.openjdk.java.net/jdk17/pull/151.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/util/random/RandomGenerator.java [hunk 1]
```diff
@@ -634,9 +640,13 @@ default int nextInt() {
...
      *
-     * @implSpec The default implementation simply calls
-     * {@link RandomSupport#checkBound checkBound}(bound) and then
-     * {@link RandomSupport#boundedNextInt boundedNextInt}(this, bound).
+     * @implSpec The default implementation checks that {@code bound} is a
+     * positive int. Then invokes {@code nextInt()}, limiting the result to
+     * be greater equal zero and less than {@code bound}. If {@code bound} is a
+     * power of two then limiting is a simple masking operation. Otherwise, a
+     * new result is re-calculated by averaging the previous result and
+     * {@code nextInt()} until the final result is greater equal zero and less
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/util/random/RandomGenerator.java:648]: I don't see how 'averaging' comes from the invocation of boundedNextInt which appears to choose the first candidate that meets the criteria.
The comment also applies to nextLong overloads.

Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/RandomGenerator.java:648]: If the first value calculated doesn't meet the criteria, then you either mask (power of 2) or go into this loop with does the averaging. The intent is not to bias the outcome.

```
                for (long u = r >>> 1;            // ensure nonnegative
                     u + m - (r = u % n) < 0L;    // rejection check
                     u = rng.nextLong() >>> 1) // retry
                    ;
```

Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/RandomGenerator.java]: Yea you're right. I was focused on the divide by two, which appears to be to just remove the sign. Clearer interpretation:

```
r = rng.nextLong();
n = range;
m = n - 1;                   // mask
for (long u = r / 2;         // ensure nonnegative
     u + m < r;              // rejection check
     u = rng.nextLong() / 2) // retry
    r = u % n;
```

File: src/java.base/share/classes/java/util/random/RandomGenerator.java [hunk 2]
```diff
@@ -657,9 +666,13 @@ default int nextInt(int bound) {
...
      *
-     * @implSpec The default implementation simply calls
-     * {@link RandomSupport#checkBound(long) checkBound}(bound) and then
-     * {@link RandomSupport#boundedNextInt(RandomGenerator, int) boundedNextInt}(this, bound).
+     * @implSpec The default implementation checks that {@code origin} and
+     * {@code bound} are positive ints. Then invokes {@code nextInt()}, limiting
+     * the result to be greater equal {@code origin} and less than {@code bound}.
+     * If {@code bound} is a power of two then limiting is a simple masking
+     * operation. Otherwise, the result is re-calculated  by invoking
+     * {@code nextInt()} until the result is greater equal {@code origin}
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/util/random/RandomGenerator.java:674]: greater equal -> greater than or equal
and in the overloads below (and above)

File: src/java.base/share/classes/java/util/random/RandomGenerator.java [hunk 3]
```diff
@@ -634,9 +640,12 @@ default int nextInt() {
      *
      * @throws IllegalArgumentException if {@code bound} is not positive
      *
-     * @implSpec The default implementation simply calls
-     * {@link RandomSupport#checkBound checkBound}(bound) and then
-     * {@link RandomSupport#boundedNextInt boundedNextInt}(this, bound).
+     * @implSpec The default implementation checks that {@code bound} is a
+     * positive int. Then invokes {@code nextInt()}, limiting the result to
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/util/random/RandomGenerator.java:644]: I think if you are refering to the Java type it should be in {@ code xxx }.
"int" - > {@ code int}.  Or use "integer".
Ditto "ints", "long", and "longs" in the overridden methods below.

File: src/java.base/share/classes/java/util/random/RandomGenerator.java (hunk 2 above)
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/RandomGenerator.java]: Updated

File: src/java.base/share/classes/java/util/random/RandomGenerator.java (hunk 3 above)
Comment (JimLaskey) [reply src/java.base/share/classes/java/util/random/RandomGenerator.java]: Updated

**Review by RogerRiggs — APPROVED** [review]
Thanks for the updates.

**Pull request closed without merging**